OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs

```diff
@@ -0,0 +1,378 @@
+				if (implicitInheritedSequences.Contains(sequenceNode.Key) && !sequenceNode.Value.Nodes.Any())
+					imageNode.RemoveNode(sequenceNode);
+
+			yield break;
+		}
+
+		void ProcessNode(ModData modData, MiniYamlNode sequenceNode, MiniYamlNode resolvedSequenceNode, string imageName)
+		{
+			var addExtension = true;
+			var addExtensionNode = resolvedSequenceNode.LastChildMatching("AddExtension");
+			if (addExtensionNode != null)
+				addExtension = FieldLoader.GetValue<bool>("AddExtension", addExtensionNode.Value.Value);
+
+			var useTilesetExtension = false;
+			var useTilesetExtensionNode = resolvedSequenceNode.LastChildMatching("UseTilesetExtension");
+			if (useTilesetExtensionNode != null)
+				useTilesetExtension = FieldLoader.GetValue<bool>("UseTilesetExtension", useTilesetExtensionNode.Value.Value);
+
+			var useTilesetCode = false;
+			var useTilesetCodeNode = resolvedSequenceNode.LastChildMatching("UseTilesetCode");
+			if (useTilesetCodeNode != null)
+				useTilesetCode = FieldLoader.GetValue<bool>("UseTilesetCode", useTilesetCodeNode.Value.Value);
+
+			var tilesetOverrides = new Dictionary<string, string>();
+			var tilesetOverridesNode = resolvedSequenceNode.LastChildMatching("TilesetOverrides");
+			if (tilesetOverridesNode != null)
+				foreach (var tilesetNode in tilesetOverridesNode.Value.Nodes)
+					tilesetOverrides[tilesetNode.Key] = tilesetNode.Value.Value;
+
+			sequenceNode.RemoveNodes("AddExtension");
+			sequenceNode.RemoveNodes("UseTilesetExtension");
+			sequenceNode.RemoveNodes("UseTilesetCode");
+			sequenceNode.RemoveNodes("TilesetOverrides");
+
+			// Replace removals with masking
+			foreach (var node in sequenceNode.Value.Nodes)
+				if (node.Key?.StartsWith("-") ?? false)
+					node.Key = node.Key.Substring(1);
+
+			var combineNode = sequenceNode.LastChildMatching("Combine");
+			if (combineNode != null)
+			{
+				var i = 0;
+				foreach (var node in combineNode.Value.Nodes)
+				{
+					ProcessNode(modData, node, node, node.Key);
+					node.Key = (i++).ToString();
+				}
+
+				return;
+			}
+
+			var filename = string.IsNullOrEmpty(resolvedSequenceNode.Value.Value) ? imageName : resolvedSequenceNode.Value.Value;
+			if (filename.StartsWith("^"))
+				return;
+
+			if (useTilesetExtension || useTilesetCode)
+			{
+				var tilesetFilenamesNode = new MiniYamlNode("TilesetFilenames", "");
+				var duplicateCount = new Dictionary<string, int>();
+				foreach (var tileset in modData.DefaultTerrainInfo.Keys)
+				{
+					if (!tilesetOverrides.TryGetValue(tileset, out var sequenceTileset))
+						sequenceTileset = tileset;
+
+					var overrideFilename = filename;
+					if (useTilesetCode)
+						overrideFilename = filename.Substring(0, 1) + tilesetCodes[sequenceTileset] +
+						                   filename.Substring(2, filename.Length - 2);
+
+					if (addExtension)
+						overrideFilename += useTilesetExtension ? tilesetExtensions[sequenceTileset] : defaultSpriteExtension;
```

Comment:
We could just check if the `sequenceTileset` exists and if it doesn't just not do the step. Instead we could add a note to the modder to fix this manually

Reply:
The same issue could happen with TilesetCodes as well correct?